### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Matrix.ZPow

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
Remaining questions / things to pay attention to:

 - [ ] Is it ok to silence the `bit0` and `bit1` deprecated warnings or should we do something else about it?
 - [ ] Do we still need `coe_units_zpow`?
 - [ ] etaExperiment in the last theorem `conjTranspose_zpow`: why is it not finding the instance `StartRing R` without it: it is just there!

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean [hunk 1]
```diff
@@ -308,19 +314,21 @@ theorem Commute.mul_zpow {A B : M} (h : Commute A B) : ∀ i : ℤ, (A * B) ^ i
...
 #align matrix.zpow_bit0' Matrix.zpow_bit0'
 
+set_option linter.deprecated false in
 theorem zpow_bit1' (A : M) (n : ℤ) : A ^ bit1 n = (A * A) ^ n * A := by
   rw [zpow_bit1, Commute.mul_zpow (Commute.refl A)]
 #align matrix.zpow_bit1' Matrix.zpow_bit1'
 
 theorem zpow_neg_mul_zpow_self (n : ℤ) {A : M} (h : IsUnit A.det) : A ^ (-n) * A ^ n = 1 := by
-  rw [zpow_neg h, mul_eq_mul, nonsing_inv_mul _ (h.det_zpow _)]
+  rw [zpow_neg h, mul_eq_mul, nonsing_inv_mul _ (IsUnit.det_zpow h _)]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Zpow.lean:331]: Does dot notation really not work here?

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean [hunk 2]
```diff
@@ -123,25 +123,25 @@ theorem zpow_neg_coe_nat (A : M) (n : ℕ) : A ^ (-n : ℤ) = (A ^ n)⁻¹ := by
 #align matrix.zpow_neg_coe_nat Matrix.zpow_neg_coe_nat
 
 theorem IsUnit.det_zpow {A : M} (h : IsUnit A.det) (n : ℤ) : IsUnit (A ^ n).det := by
-  cases n
+  cases' n with n n
   · simpa using h.pow n
   · simpa using h.pow n.succ
-#align is_unit.det_zpow IsUnit.det_zpow
+#align is_unit.det_zpow Matrix.IsUnit.det_zpow
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Zpow.lean:129]: This change is incorrect. You should almost never edit an `#align`. You need to fix the name of the lemma instead.

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Zpow.lean:1]: Should this be `ZPow`?

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean [hunk 3]
```diff
@@ -285,19 +287,23 @@ theorem zpow_mul' (A : M) (h : IsUnit A.det) (m n : ℤ) : A ^ (m * n) = (A ^ n)
   rw [mul_comm, zpow_mul _ h]
 #align matrix.zpow_mul' Matrix.zpow_mul'
 
+/-
+porting note: TODO: is this still needed, or should just the norm_cast be removed?
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Zpow.lean:291]: Lemmas should almost never be removed during porting. Why does `norm_cast` complain?

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean [hunk 4]
```diff
@@ -285,19 +287,23 @@ theorem zpow_mul' (A : M) (h : IsUnit A.det) (m n : ℤ) : A ^ (m * n) = (A ^ n)
...
+-/
+#noalign matrix.coe_units_zpow
 
 theorem zpow_ne_zero_of_isUnit_det [Nonempty n'] [Nontrivial R] {A : M} (ha : IsUnit A.det)
     (z : ℤ) : A ^ z ≠ 0 := by
-  have := ha.det_zpow z
-  contrapose! this
-  rw [this, det_zero ‹_›]
-  exact not_isUnit_zero
+  have := IsUnit.det_zpow ha z
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/Zpow.lean:303]: This change can be reverted, it's due to the `#align` above not being fixed properly

**Comment by eric-wieser** [comment]
> * [ ]  Is it ok to silence the `bit0` and `bit1` deprecated warnings or should we do something else about it?

I think this is the right approach, though maybe the lemmas themselves should be marked deprecated; have a look at how we handle similar things elsewhere

> * [ ]  Do we still need `coe_units_zpow`?

Yes, why wouldn't we?

> * [ ]  etaExperiment in the last theorem `conjTranspose_zpow`: why is it not finding the instance `StartRing R` without it: it is just there!

Because the implicit arguments to `StarRing` are different. This is expected behavior.

> * [ ]  the name of the file should stay `Zpow` and not `ZPow`, right?

I'd lean towards `ZPow` but perhaps just ask on Zulip.

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean (hunk 3 above)
Comment (mo271) [reply Mathlib/LinearAlgebra/Matrix/Zpow.lean]: on the left hand side, `u ^ n` was interpreted in the wrong way, I think it should be `(u : M) ^ n`. Now it works

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean (hunk 4 above)
Comment (mo271) [reply Mathlib/LinearAlgebra/Matrix/Zpow.lean]: yes, except that 
`contrapose! this`  still does not work

File: Mathlib/LinearAlgebra/Matrix/Zpow.lean (hunk 1 above)
Comment (mo271) [reply Mathlib/LinearAlgebra/Matrix/Zpow.lean]: After changing the lemma above it does now work in all places.

**Comment by mo271** [comment]
> > * [ ]   Is it ok to silence the `bit0` and `bit1` deprecated warnings or should we do something else about it?
> 
> I think this is the right approach, though maybe the lemmas themselves should be marked deprecated; have a look at how we handle similar things elsewhere

Seems like we do the same, i.e. just disabling the warnings, in other places and don't mark them as deprecated in any other way.

> 
> > * [ ]   Do we still need `coe_units_zpow`?
> 
> Yes, why wouldn't we?
This is done now.

> 
> > * [ ]   the name of the file should stay `Zpow` and not `ZPow`, right?
> 
> I'd lean towards `ZPow` but perhaps just ask on Zulip.

I'll ask on Zulip. There is already `Mathlib.GroupTheory.Subgroup.Zpowers`, so if we aim for consistency we might want to change both?!

**Comment by mo271** [comment]
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4

**Comment by semorrison** [comment]
bors merge

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**